include/swift/Parse/ParsedRawSyntaxNode.h

```diff
@@ -150,6 +102,24 @@ class ParsedRawSyntaxNode {
     assert(ensureDataIsNotRecorded() && "recorded data is being destructed");
   }
 
+  /// Returns the type of this node (recorded, deferred layout, deferred token,
+  /// null).
+  DataKind getNodeType() const { return Data.getKind(); }
+
+  /// Returns the opaque data of this node. This must be interpreted by the
+  /// \c SyntaxParseAction, which likely also needs the node type to know
+  /// what type of node the data represents.
+  OpaqueSyntaxNode getData() const { return Data.getOpaque(); }
+
+  RecordedOrDeferredNode getRecordedOrDeferredNode() const { return Data; }
```

Comment:
These methods (`getData()` and `getRecordedOrDeferredNode()`) breaks the "move-only" semantics of `ParsedRawSyntaxNode`. "Data" should not be extracted without resetting the `ParsedRawSyntaxNode`.
This is important to ensure not to discard parsed recorded nodes without putting them into the final tree.

Previously, we allowed copying via `copyDeferred()` method only if it's deferred.

Reply:
We need some kind of “get data, but don’t take” method to retrieve the node’s children [here](https://github.com/apple/swift/pull/36277/files#diff-a83cf4327ae0cebc7c22c773cfa0aac6304f414828248800d7492f1e3c4145a8R214).

I’ve renamed the method to `getUnsafeOpaqueData` and added a comment that the data returned by it should only be used to inspect the node, not to record it. Are you OK with that?

Also, while looking through the uses of this method, I actually found a case where we should have been using `takeData` instead, so good catch.